prototype-spec.md

```diff
@@ -0,0 +1,253 @@
+terminating delegations instruct the client not to consider future trust
+statements that match the delegation's pattern, which stops the delegation
+processing once this delegation (and its descendants) have been processed.
+A terminating delegation for a package causes any further statements about a
+package that are not made by the delegated party or its descendants to be
+ignored.
+
+* In order to discuss target paths, a role MUST specify only one of the "path_hash_prefixes", "succinct_hash_delegations", or "paths" attributes, each of which we discuss next.
+   * `path_hash_prefixes` :: A list of HEX_DIGESTs used to succinctly describe a set of target paths. Specifically, each HEX_DIGEST in "path_hash_prefixes" describes a set of target paths; therefore, "path_hash_prefixes" is the union over each prefix of its set of target paths. The target paths must meet this condition: each target path, when hashed with the SHA-256 hash function to produce a 64-byte hexadecimal digest (HEX_DIGEST), must share the same prefix as one of the prefixes in "path_hash_prefixes". This is useful to split a large number of targets into separate bins identified by consistent hashing.
+   * `succinct_hash_delegations` :: A dictionary containing succinct descriptions of delegations to a set of bins, similar to those used in `path_hash_prefixes`. This field represents delegations to 2^BIT_LENGTH bins that use the keyid, threshold, path, and termination status from this delegation. The path_hash_prefixes and name for each bin will be determined using the BIT_LENGTH and NAME_PREFIX. BIT_LENGTH must be an integer between 1 and 32 (inclusive). The rolename for each bin will be structured as NAME_PREFIX-COUNT where COUNT is a hexadecimal value between 0 and 2^BIT_LENGTH-1 (inclusive) that represents the bin number. If the succinct_hash_delegations field is present in a delegation, the name field will not be used as a rolename, and so is not required.
+   * `paths` :: A list of strings, where each string describes a path that the role is trusted to provide. Clients MUST check that a target is in one of the trusted paths of all roles in a delegation chain, not just in a trusted path of the role that describes the target file. PATHPATTERN can include shell-style wildcards and supports the Unix filename pattern matching convention. Its format may either indicate a path to a single file, or to multiple paths with the use of shell-style wildcards. For example, the path pattern "targets/*.tgz" would match file paths "targets/foo.tgz" and "targets/bar.tgz", but not "targets/foo.txt". Likewise, path pattern "foo-version-?.tgz" matches "foo-version-2.tgz" and "foo-version-a.tgz", but not "foo-version-alpha.tgz". To avoid surprising behavior when matching targets with PATHPATTERN, it is RECOMMENDED that PATHPATTERN uses the forward slash (/) as directory separator and does not start with a directory separator, akin to TARGETPATH.
```

Comment:
To avoid surprising behavior when matching targets with PATHPATTERN, it is RECOMMENDED that PATHPATTERN uses the forward slash (/) as a directory separator, but does not start with a directory separator, akin to TARGETPATH.